library/blocks/block-extensions/src/main/java/org/quiltmc/qsl/block/extensions/api/data/BlockDataKey.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.block.extensions.api.data;
+
+/**
+ * Represents a key that can be used to get a value from an {@link ExtraBlockData} instance.
+ *
+ */
+@SuppressWarnings("ClassCanBeRecord") // this key is identity-based, so it can't be a record!
+public final class BlockDataKey<T> {
```

Comment:
Bit worried about the identity-baseness of this class as you will need to get a hold of the key reference for your stuff.

Like if you had something where the type is Float, which is managed by an external library, you're forced to bring that library in your dev env to be able to use the key since it's identity-based.

Reply:
This entire system is being replaced (as soon as I get back to my laptop and also sleep), so this PR will be on hiatus until the new system is done, which will be another QSL library.